File: MATTHEW/leetcode/AIcoding/2.Gemini-2.0/cpp_solutions/655-print-binary-tree.cpp
#include <vector>
#include <queue>
#include <cmath>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<vector<string>> printTree(TreeNode* root) {
        int height = getHeight(root);
        int width = pow(2, height) - 1;
        vector<vector<string>> result(height, vector<string>(width, ""));
        fillTree(root, result, 0, 0, width - 1);
        return result;
    }

private:
    int getHeight(TreeNode* root) {
        if (!root) return 0;
        return 1 + max(getHeight(root->left), getHeight(root->right));
    }

    void fillTree(TreeNode* root, vector<vector<string>>& result, int row, int left, int right) {
        if (!root) return;
        int mid = (left + right) / 2;
        result[row][mid] = to_string(root->val);
        fillTree(root->left, result, row + 1, left, mid - 1);
        fillTree(root->right, result, row + 1, mid + 1, right);
    }
};

File: leetcode/AIcoding/2.Gemini-2.0/java_solutions/655-print-binary-tree.java
class Solution {
    public List<List<String>> printTree(TreeNode root) {
        int height = getHeight(root);
        int width = (int)Math.pow(2, height) - 1;
        List<List<String>> result = new ArrayList<>();
        for (int i = 0; i < height; i++) {
            List<String> row = new ArrayList<>();
            for (int j = 0; j < width; j++) {
                row.add("");
            }
            result.add(row);
        }
        populate(root, result, 0, 0, width - 1);
        return result;
    }

    private int getHeight(TreeNode root) {
        if (root == null) {
            return 0;
        }
        return 1 + Math.max(getHeight(root.left), getHeight(root.right));
    }

    private void populate(TreeNode root, List<List<String>> result, int row, int left, int right) {
        if (root == null) {
            return;
        }
        int mid = (left + right) / 2;
        result.get(row).set(mid, String.valueOf(root.val));
        populate(root.left, result, row + 1, left, mid - 1);
        populate(root.right, result, row + 1, mid + 1, right);
    }
}